boot/model_test.go

```diff
@@ -1067,3 +1083,158 @@ func (s *modelSuite) TestDeviceChangeUnhappyFailReseaWithSwappedModelMockedWrite
 	// try model has been cleared
 	c.Assert(tryForSealing, Equals, "/,,")
 }
+
+func (s *modelSuite) TestDeviceChangeRebootRestoreModelKeyChangeMockedWriteModel(c *C) {
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	oldKeyID := "Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij"
+	newKeyID := "ZZZ_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij"
+	// model can be mocked freely as we will not encode it
+	oldUC20dev := boottest.MockUC20Device("", boottest.MakeMockUC20Model(map[string]interface{}{
+		"model":             "my-model-uc20",
+		"brand-id":          "my-brand",
+		"grade":             "dangerous",
+		"sign-key-sha3-384": oldKeyID,
+	}))
+
+	newUC20dev := boottest.MockUC20Device("", boottest.MakeMockUC20Model(map[string]interface{}{
+		"model":             "my-model-uc20",
+		"brand-id":          "my-brand",
+		"grade":             "dangerous",
+		"sign-key-sha3-384": newKeyID,
+	}))
+
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		s.readSystemEssentialCalls++
+		kernelRev := 1
+		systemModel := oldUC20dev.Model()
+		if label == "1234" {
+			// recovery system for new model
+			kernelRev = 999
+			systemModel = newUC20dev.Model()
+		}
+		return systemModel, []*seed.Snap{mockKernelSeedSnap(c, snap.R(kernelRev)), mockGadgetSeedSnap(c, nil)}, nil
+	})
+	defer restore()
+
+	resealKeysCalls := 0
+	restore = boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error {
+		resealKeysCalls++
+		c.Logf("reseal key call: %v", resealKeysCalls)
+		m, err := boot.ReadModeenv("")
+		c.Assert(err, IsNil)
+		currForSealing := boot.ModelUniqueID(m.ModelForSealing())
+		tryForSealing := boot.ModelUniqueID(m.TryModelForSealing())
+		// timeline & calls:
+		// 1, 2 - pre reboot, run & recovery keys, try model set
+		// 3 - run key, after model file has been modified, try model cleared
+		// 4 - recovery key, model file has been modified, triggers reboot, try model cleared
```

Comment:
I didn't think of it before,  but triggers reboot make it sound like it's expected, instead we are simulating unexpected reboots, no? so maybe everywhere

triggers reboot/triggers 'reboot'  ->  unexpected reboot is triggered